### DISCUSSION TITLE
Get selected text from Label

### DISCUSSION THREAD

**Discussion by eleanor-em** [emilk/egui#4913] [Q&A] [answered]
This seems simple, but I can't find any information on this: suppose I have a `Label`:
```rs
ui.add(egui::Label::new("here is some text")
    .selectable(true));
```
How do I get the text that has been selected from this label? The returned `Response` object does not seem to have anything useful in it, unless I can somehow use another API to "look up" the label from the `Id` or similar. (The goal here is to support copying selected text to the clipboard.)

**Comment by rustbasic** [comment]
If you do `selectable(true)`, you can select the contents of the label with the mouse.
And you can copy it with Ctrl + C (in Windows).

Since this is a new feature, it seems that there will be additional features in the future.

**Reply by eleanor-em** [reply]
Thanks! You're right that copying works with `selectable(true)`, I was not correctly applying the `platform_output` when I tried it.

However I'd still like to know if it is possible to access the selected text more directly. It seems you are saying it isn't?

**Comment by YgorSouza** [comment, marked as answer]
As rustbasic said, the selected text is already copied to the clipboard when a Copy event is received. I don't think it is possible to copy or access the selected text under different circumstances. The text to be copied is stored in this private field:

https://github.com/emilk/egui/blob/d856f7b3a58fc65a81ff9657f5c6e1ab462b7bb6/crates/egui/src/text_selection/label_text_selection.rs#L122-L123

This field is normally empty, as it is not needed on every frame. When a copy event arrives, every label appends its selected text to the accumulated selection as it is painted:

https://github.com/emilk/egui/blob/d856f7b3a58fc65a81ff9657f5c6e1ab462b7bb6/crates/egui/src/text_selection/label_text_selection.rs#L510-L512

Then, at the end of the frame, the accumulated text is moved to the clipboard (using a public method that you can also call from your code):

https://github.com/emilk/egui/blob/d856f7b3a58fc65a81ff9657f5c6e1ab462b7bb6/crates/egui/src/text_selection/label_text_selection.rs#L228-L231

So you can put arbitrary strings in the clipboard using [copy_text](https://docs.rs/egui/0.28.1/egui/struct.Context.html#method.copy_text), and egui will put the selected text in the clipboard automatically when the user presses Ctrl+C (or the Copy shortcut in their system), but you can't retrieve the selected text to put it in the clipboard yourself.

On the other hand, you can use a `TextEdit` with a non-mut string (which makes it look like a label), and then you can access the selected text like this:

```rs
#[derive(Default)]
struct MyApp {}

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            use egui::TextBuffer as _;
            let text = "The quick brown fox jumps over the lazy dog";
            let response = TextEdit::multiline(&mut text.as_str()).show(ui);
            let selection = response
                .cursor_range
                .as_ref()
                .map(|c| text.char_range(c.as_sorted_char_range()));
            if let Some(selection) = selection {
                ui.label(selection);
            }
        });
    }
}
```

![image](https://github.com/user-attachments/assets/64a3bf99-7a67-4ca8-9a03-d2def9a054eb)


But if you plan on copying it to the clipboard when the user clicks a button on something similar, be aware that clicking the button will clear the selection before you have a chance to copy it to the clipboard, so you have to store it and get the selection from a previous frame when the button is clicked.

**Reply by eleanor-em** [reply]
Thank you very much for the detailed response!
